api/desecapi/permissions.py

```diff
@@ -13,6 +13,49 @@ def has_permission(self, request, view):
         return request.user and request.user.is_active
 
 
+class IsAPIToken(permissions.BasePermission):
+    """
+    Allows access only with API token (.mfa is None).
+    """
+
+    message = "API token required."
+    code = "api_token_required"
+
+    def has_permission(self, request, view):
+        return request.auth.mfa is None
+
+
+class IsLoginToken(permissions.BasePermission):
+    """
+    Allows access only with login token (.mfa is not None).
+
+    DRF permission negation is flawed, so ~IsAPIToken does not give the correct behavior:
+    https://github.com/encode/django-rest-framework/issues/6598#issuecomment-484824743
+    """
+
+    message = "Login token required."
+    code = "login_token_required"
+
+    def has_permission(self, request, view):
+        return request.auth.mfa is not None
+
+
+class MFAIfEnabled(permissions.BasePermission):
+    """
+    Allows access only to when
+        - the token is a human token that has passed MFA, or
+        - the token is a human token that has not passed MFA, but the user has not enabled MFA at all.
+    """
+
+    message = "Multi-factor authentication required."
+    code = "mfa_required"
+
+    def has_permission(self, request, view):
+        return request.auth.mfa or (
+            request.auth.mfa is False and not request.user.mfa_enabled
+        )
```

Comment:
> Allowing more, we could say: `request.auth.mfa or not request.use.mfa_enabled`

I'm not sure if you're proposing to replace the full 3-part condition with this one. But if we do that, the consequence would be that API tokens (`.mfa=None`) fulfill this condition as long as MFA has not been set up. Then, once MFA is set up, API tokens no longer fulfill the condition, so by enabling MFA, the set up endpoints that can be accessed using an API token is changed silently.

I think we shouldn't do that. Whether MFA is enabled or not is an attribute of a user account, not of a token. Each token should, in principle, be applicable to a static set of endpoints, regardless of whether MFA is enabled or not (modulo sometimes requiring step-up). Hope that makes sense.

> Let's say a user disables MFA by removing all factors, should this invalidate the issued tokens?

This needs distinction between log-in tokens and API tokens.

- **Log-in tokens.** We discussed that last week on the phone, and I was in favor of that behavior. (I said that if we link each token with the factor used for step-up via a foreign key, then the token could be deleted by the dmbs upon factor deletion.) We ended up not going this route, so for now the behavior is that tokens remain valid when all MFA factors are removed.
  I'm open to changing that, though, because I still think it's a good idea to reset all user sessions whenever *any* single authentication credential is invalidated (e.g.: invalidate user sessions at password change, MFA factor removal, ...). 
  Related (but not the same): https://medium.com/@lukeberner/how-i-abused-2fa-to-maintain-persistence-after-a-password-change-google-microsoft-instagram-7e3f455b71a1 I believe there was also an earlier vulnerability. -- In any case, I think caution is advised. Users usually won't have multiple sessions, so usually the invalidation of sessions will amount to no-op. And if there are, then it may be wise to assume that they *may* no longer be legit.

- **API tokens.** Per the same argument(s), one may say that API tokens should be invalidated when credentials change. I think that is too intrusive, but perhaps the GUI should offer a "Delete all API tokens" button whenever a credential change is performed. -- A less intrusive way may be to offer deletion of tokens since the previous credential change (we have this as a user attribute now). -- I think this aspect doesn't need to be handled, now, though. Do you think it's worth an issue? (I would think we'd arrive at doing nothing by default, and perhaps only offering the button.)

Reply:
## `MFARequiredIfEnabled.has_permission`

### Current State


|                          | `token.mfa=None` (API token)   | `token.mfa=False`   | `token.mfa=True`   |
|:-------------------------|:-------------------------------|:--------------------|:-------------------|
| `user.mfa_enabled=True`  | False                          | False               | True               |
| `user.mfa_enabled=False` | False                          | True                | True               |

- Non-API tokens not have it if and only if they are not stepped up and the user's MFA is enabled.
- API tokens never have `MFARequiredIfEnabled` permission.

This means that Non-API tokens that are not stepped up (`token.mfa=False`) can toggle between having permission and not having permission when the user's MFA is en/disabled.

### Relaxed Proposal

|                          | `token.mfa=None` (API token)   | `token.mfa=False`   | `token.mfa=True`   |
|:-------------------------|:-------------------------------|:--------------------|:-------------------|
| `user.mfa_enabled=True`  | False                          | False               | True               |
| `user.mfa_enabled=False` | True                           | True                | True               |

- Non-API tokens not have it if and only if they are not stepped up and the user's MFA is enabled. (Same as above)
- API tokens have permission if the user's MFA is not enabled.

This means that Non-API tokens *and* API tokens can toggle between having permission and not having permission when the user's MFA is en/disabled.

### Comparison

In both versions, token permissions can toggle. However, we expect Non-API tokens to live shorter than API tokens, the current state will effectively toggle less.

Nevertheless, I find this a bit convoluted as the permission is `False` for API-tokens, but exclusively used in "OR"-combination with a permission that is `True` for API tokens. That makes the permissions for views hard to read (to me anyways).

Also, I'm still skeptical about the distinction we make here with API tokens and login tokens based on the (implicit) MFA parameter. By above argument, isn't the short/longevity of the token the argument why one is subject to MFA while the other isn't? (Should we enforce short validity for login tokens?)

Reply:
Decided offline to keep the current state. Permissions are fine the way they are currently but may need an overhaul in the long term.